### ISSUE TITLE
[Spike] Investigate using storybook for E2E testing

### ISSUE THREAD

**Issue by Gkrumbach07** [opendatahub-io/odh-dashboard#1043]
### Feature description

I want to sue storybook to e2e test the dashboard. We need to discover what needs to work and what needs to be done in order fort storybook to be viable for e2e.

The main considerations for storybook to be viable is:
- consistency
- reusability

**Consistency**
If we cant guarantee the backend is in a certain state each time we run a story, then this wont work. There are are ways to globally and per story run setup and tear functions, so i am confident about this. 

**Reusability**
If we cant reuse our mocked stories, then rewriting the same story just for e2e will be pointless and we might as well use something that is built for 2e2. But if we can use the mocked tests then e2e is essentially free. For example, we created a mocked test for a form page to check if the form behaves correctly all the way up till submission. Then we want to e2e test this by actually submitting. can we reuse the story we created so we dont have to recode the steps to fill the form?

### Describe alternatives you've considered

We could use cyprus, which seems like what most people recommend (even storybook mentions cyprus albeit an outdated doc). This is somewhat new territory as storybook is delving into the relm where it could in theory do e2e tests but people have not really tried yet.

### Anything else?

This will be a spike to investigate if this is possible

**Comment by lucferbux** [comment]
I have  some thoughts about this hole process, I'm just gonna add some definitions in case I'm wrong about the premise.
Currently, we are aiming for these three types of tests:

* **unit testing**: These tests are meant to run locally, they cover small chunks of code, and they are fairly useful to maintain all the `utils` files that we have.
* **integration testing:**: These cover all the integration between  chunks of code, in case of react integration is useful to  test components and how they interact with each other. 

For these two kind of tests, having mock data is very useful, cause we can test things like `state`, `props` and more about all the components we create. I think we are really fine right now on how Storybook handles this and I think we are on the right track here. We can argue that UI testing might fall back into integration testing if we are testing clicks and inputs and so, but it's fine for me.

* **e2e**:  Verify the code as the end user would use it. This means that we can't leverage testing on mocks and so the scope is broader than the  last 2 pieces.

In my opinion, we already have support for e2e validation with `Playwright`, the goal there is just to have testing with a headless browser and we are already doing this. The **main challenge** here is to have a dedicated environment with the latest version of the operator to get this kind of testing enabled, I think it's the only way to really achieve a good e2e testing coverage.

**Comment by lucferbux** [comment]
> I have some thoughts about this hole process, I'm just gonna add some definitions in case I'm wrong about the premise. Currently, we are aiming for these three types of tests:
> 
> * **unit testing**: These tests are meant to run locally, they cover small chunks of code, and they are fairly useful to maintain all the `utils` files that we have.
> * **integration testing:**: These cover all the integration between  chunks of code, in case of react integration is useful to  test components and how they interact with each other.
> 
> For these two kind of tests, having mock data is very useful, cause we can test things like `state`, `props` and more about all the components we create. I think we are really fine right now on how Storybook handles this and I think we are on the right track here. We can argue that UI testing might fall back into integration testing if we are testing clicks and inputs and so, but it's fine for me.
> 
> * **e2e**:  Verify the code as the end user would use it. This means that we can't leverage testing on mocks and so the scope is broader than the  last 2 pieces.
> 
> In my opinion, we already have support for e2e validation with `Playwright`, the goal there is just to have testing with a headless browser and we are already doing this. The **main challenge** here is to have a dedicated environment with the latest version of the operator to get this kind of testing enabled, I think it's the only way to really achieve a good e2e testing coverage.

So my point here is that we are mostly fine with the current set up, and we can leverage most of this testing to Storybook and Playwright, but it's imperative the use of a real cluster with openshift and the opendatahub operator enabled.

**Comment by lucferbux** [comment]
> I have some thoughts about this hole process, I'm just gonna add some definitions in case I'm wrong about the premise. Currently, we are aiming for these three types of tests:
> 
> * **unit testing**: These tests are meant to run locally, they cover small chunks of code, and they are fairly useful to maintain all the `utils` files that we have.
> * **integration testing:**: These cover all the integration between  chunks of code, in case of react integration is useful to  test components and how they interact with each other.
> 
> For these two kind of tests, having mock data is very useful, cause we can test things like `state`, `props` and more about all the components we create. I think we are really fine right now on how Storybook handles this and I think we are on the right track here. We can argue that UI testing might fall back into integration testing if we are testing clicks and inputs and so, but it's fine for me.
> 
> * **e2e**:  Verify the code as the end user would use it. This means that we can't leverage testing on mocks and so the scope is broader than the  last 2 pieces.
> 
> In my opinion, we already have support for e2e validation with `Playwright`, the goal there is just to have testing with a headless browser and we are already doing this. The **main challenge** here is to have a dedicated environment with the latest version of the operator to get this kind of testing enabled, I think it's the only way to really achieve a good e2e testing coverage.

[UPDATE 2]

The other point I wanted to make is that integration tests should be targeted to components, creating tests around integration, but e2e should be created around User Stories, meaning that these kinds of tests should be broader and they should mimic user flows, such as "Create a new project, and then create a workbench" or something like that.

**Comment by Gkrumbach07** [comment]
@lucferbux I agree with your defs and that we need separate e2e user stories. I do still see an opportunity to reuse our integration test stories (see [this](https://storybook.js.org/docs/react/writing-tests/importing-stories-in-tests#example-with-playwright)).

If we can make more general stories such as, filled form, create project, delete project. Maybe we can reuse these in playwright. 

but we will still have a separate area for these e2e tests. as they will span multiple pages
